lib/auth/accountrecovery_test.go

```diff
@@ -111,3 +133,264 @@ func TestRecoveryCodeEventsEmitted(t *testing.T) {
 	require.Equal(t, events.RecoveryCodeUsedEvent, event.GetType())
 	require.Equal(t, events.RecoveryCodeUseFailureCode, event.GetCode())
 }
+
+func TestStartAccountRecovery(t *testing.T) {
+	srv := newTestTLSServer(t)
+	ctx := context.Background()
+	fakeClock := srv.Clock().(clockwork.FakeClock)
+	mockEmitter := &events.MockEmitter{}
+	srv.Auth().emitter = mockEmitter
+
+	defaultModules := modules.GetModules()
+	defer modules.SetModules(defaultModules)
+	modules.SetModules(&testWithCloudModules{})
+
+	u, err := createUserWithSecondFactorAndRecoveryCodes(srv, "otp")
+	require.NoError(t, err)
+
+	// Test with recover type 2FA.
+	startToken, err := srv.Auth().StartAccountRecovery(ctx, &proto.StartAccountRecoveryRequest{
+		Username:     u.username,
+		RecoveryCode: []byte(u.recoveryCodes[0]),
+		RecoverType:  types.UserTokenUsage_RECOVER_2FA,
+	})
+	require.NoError(t, err)
+	require.Equal(t, UserTokenTypeRecoveryStart, startToken.GetSubKind())
+	require.Equal(t, types.UserTokenUsage_RECOVER_2FA, startToken.GetUsage())
+	require.Equal(t, startToken.GetURL(), fmt.Sprintf("https://<proxyhost>:3080/web/recovery/steps/%s/verify", startToken.GetName()))
+
+	// Test token returned correct byte length.
+	bytes, err := hex.DecodeString(startToken.GetName())
+	require.NoError(t, err)
+	require.Len(t, bytes, RecoveryTokenLenBytes)
+
+	// Test expired token.
+	fakeClock.Advance(defaults.RecoveryStartTokenTTL)
+	_, err = srv.Auth().GetUserToken(ctx, startToken.GetName())
+	require.True(t, trace.IsNotFound(err))
+
+	// Test events emitted.
+	event := mockEmitter.LastEvent()
+	require.Equal(t, event.GetType(), events.RecoveryTokenCreateEvent)
+	require.Equal(t, event.GetCode(), events.RecoveryTokenCreateCode)
+	require.Equal(t, event.(*apievents.UserTokenCreate).Name, u.username)
+
+	// Test with recover type PWD.
+	startToken, err = srv.Auth().StartAccountRecovery(ctx, &proto.StartAccountRecoveryRequest{
```

Comment:
I think we can refactor this to be a table test with 2 entries (recover with MFA / recover with password), with the empty recover type either before or after the table.